base/channels.jl

```diff
@@ -457,3 +457,24 @@ function iterate(c::Channel, state=nothing)
 end
 
 IteratorSize(::Type{<:Channel}) = SizeUnknown()
+
+"""
+    threaded_foreach(f, channel::Channel; ntasks=Threads.nthreads(), async=true)
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` Tasks spawned by `Threads.@spawn`.
+
+If `async` is `false`, this function will `wait` for all internally spawned tasks
+to complete before returning.
+"""
+function threaded_foreach(f, channel::Channel; ntasks=Threads.nthreads(), async=true)
+    tasks = map(1:ntasks) do _
+        Threads.@spawn begin
+            for item in channel
+                wait(Threads.@spawn f(item))
+            end
+        end
+    end
+    async || foreach(wait, tasks)
```

Comment:
I highly appreciate if `async` option is removed.  There is discussion structured concurrency #33248 and adding "unstructured" construct such as `async=true` here makes implementing structured concurrency in Julia much harder.  You can still do `@async threaded_foreach(f, channel)` (preferably within a `@sync` block).